### PR TITLE
Add flag-controlled sliced MLA projections

### PR DESCRIPTION
**Pull request by dandragona**
Performance optimizations that avoids an unnecessary split operation.

- Add use_sliced_mla_proj flag to base.yml and types.py.
- Support slice_bounds in DenseGeneral.__call__ to slice projection weights/bias prior to contraction for unquantized paths.
- Update mla_query_projection and mla_get_key_value in attention_mla.py to use sliced projections when enabled.
- Add unit test coverage in linears_test.py.

# Tests

- added unit tests
- https://xmanager.corp.google.com/experiments/271579010
- Before: https://xprof.corp.google.com/trace_viewer/chengnuojin-11765148423170897263?view_start=1400.245&view_end=1420.311
- After: https://xprof.corp.google.com/trace_viewer/dandragona-1224952786495301116?view_start=2691.354&view_end=2701.054
- After: https://xprof.corp.google.com/overview_page/dandragona-7513819543584336902 (isolation microbench)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

CHANGE_STEWARD=true

### CODE REVIEW

File: src/maxtext/layers/linears.py [hunk 1]
```diff
@@ -279,6 +289,11 @@ def __call__(self, inputs: Array, _initializing: bool = False, out_sharding: Nam
         kernel = jax.device_put(kernel, max_utils.device_space())
       kernel = jnp.asarray(kernel, self.dtype)
 
+    if slice_bounds is not None:
+      assert self.quant is None, "sliced contraction is only supported when quant is None"
```
Comment (NuojCheng) [inline src/maxtext/layers/linears.py]: we discourage using assertion error in general, since they can be skipped sometimes. We could use `if ... raise ValueError()` instead

File: tests/unit/attention_test.py [hunk 2]
```diff
@@ -2023,6 +2023,98 @@ def test_indexer_autoregression(self, prefill_len, target_len):
...
+    out_sliced_train, _ = mla_sliced(
+        lnx,
+        lnx,
+        decoder_segment_ids=decoder_segment_ids,
+        inputs_positions=decoder_positions,
+        deterministic=True,
+        model_mode=MODEL_MODE_TRAIN,
+    )
+
+    self.assertTrue(jnp.allclose(out_normal_train, out_sliced_train, rtol=1e-05, atol=1e-05, equal_nan=False))
```
Comment (NuojCheng) [inline tests/unit/attention_test.py:2132]: for training we want to compare gradients to protect backward pass correctness

File: src/maxtext/layers/attention_mla.py [hunk 3]
```diff
@@ -917,17 +917,30 @@ def mla_query_projection(
...
       q = self.query(inputs_q, out_sharding=query_sharding)
+      q_nope, q_pe = jnp.split(q, [self.qk_nope_head_dim], axis=-1)
     else:
       # LoRA path
       low_rank_q = self.wq_a(inputs_q, out_sharding=wqa_out_sharding)  # [B, L, q_lora_rank]
       low_rank_q = checkpoint_name(low_rank_q, "query_wa_proj")
       low_rank_q = self.q_norm(low_rank_q)  # RMSNorm on low rank
       low_rank_q = checkpoint_name(low_rank_q, "mla_q")
-      q = self.wq_b(low_rank_q, out_sharding=query_sharding)  # [B, L, n_heads, qk_head_dim]
+      if self.config.use_sliced_mla_projections and self.wq_b.quant is None:
```
Comment (RissyRan) [inline src/maxtext/layers/attention_mla.py]: Shall we add assertion in types.py file to indicate that `use_sliced_mla_projections` doesn't work with `quantization` config for early check?

File: tests/end_to_end/tpu/deepseek/Run_DeepSeek.md [hunk 4]
```diff
@@ -304,6 +304,14 @@ python3 -m tests.utils.forward_pass_logit_checker \
 
 To run MMLU benchmarks and validate the model's performance, follow the instructions provided [here](https://github.com/AI-Hypercomputer/maxtext/blob/main/benchmarks/api_server/README.md).
 
+## MLA Optimization
+
+To optimize Multi-Head Latent Attention (MLA) performance, you can enable sliced projections.
+
+*   **Flag**: `use_sliced_mla_projections` (default: `False`)
```
Comment (RissyRan) [inline tests/end_to_end/tpu/deepseek/Run_DeepSeek.md]: I see we put `proj` as short name in base.yml. Maybe we could make it as `use_sliced_mla_proj`?

**Review by RissyRan — APPROVED** [review]
Thanks! LGTM, please address comments before merge.

File: src/maxtext/layers/linears.py [hunk 5]
```diff
@@ -279,6 +289,12 @@ def __call__(self, inputs: Array, _initializing: bool = False, out_sharding: Nam
         kernel = jax.device_put(kernel, max_utils.device_space())
       kernel = jnp.asarray(kernel, self.dtype)
 
+    if slice_bounds is not None:
+      if self.quant is not None:
+        raise ValueError("sliced contraction is only supported when quant is None")
+      begin, end = slice_bounds
+      kernel = kernel[..., begin:end]
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply src/maxtext/layers/linears.py:300]: Please consider.

File: tests/end_to_end/tpu/deepseek/Run_DeepSeek.md (hunk 4 above)
Comment (dandragona) [reply tests/end_to_end/tpu/deepseek/Run_DeepSeek.md]: Done.

File: src/maxtext/layers/attention_mla.py (hunk 3 above)
Comment (dandragona) [reply src/maxtext/layers/attention_mla.py]: Done.

File: src/maxtext/layers/linears.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dandragona) [reply src/maxtext/layers/linears.py:300]: Done.

**Review by NuojCheng — APPROVED** [review]
